File: macos/Tests/NoctusCoreTests/KeyBindingResolverTests.swift
import Testing
@testable import NoctusCore

@Suite struct KeyBindingResolverTests {

    @Test func resolveRegisteredBinding() {
        let r = KeyBindingResolver()
        r.setBinding("pane.copy", chord: KeyChord("F5"))
        #expect(r.resolve(KeyChord("F5")) == "pane.copy")
    }

    @Test func resolveUnknownReturnsNil() {
        let r = KeyBindingResolver()
        #expect(r.resolve(KeyChord("F12")) == nil)
    }

    @Test func setBindingOverrides() {
        let r = KeyBindingResolver()
        r.setBinding("old", chord: KeyChord("F5"))
        r.setBinding("new", chord: KeyChord("F5"))
        #expect(r.resolve(KeyChord("F5")) == "new")
    }

    @Test func getBindingReturnsChord() {
        let r = KeyBindingResolver()
        let chord = KeyChord("P", ctrl: true, shift: true)
        r.setBinding("cmd.palette", chord: chord)
        #expect(r.getBinding("cmd.palette") == chord)
    }

    @Test func removeBindingWorks() {
        let r = KeyBindingResolver()
        r.setBinding("cmd", chord: KeyChord("F5"))
        r.removeBinding("cmd")
        #expect(r.resolve(KeyChord("F5")) == nil)
        #expect(r.getBinding("cmd") == nil)
    }

    @Test func loadBindingsClearsPrevious() {
        let r = KeyBindingResolver()
        r.setBinding("old", chord: KeyChord("F1"))
        r.loadBindings(["new": KeyChord("F2")])
        #expect(r.resolve(KeyChord("F1")) == nil)
        #expect(r.resolve(KeyChord("F2")) == "new")
    }
}

File: tests/Explorer.Core.Tests/Services/KeyBindingResolverTests.cs
using NoctusExplorer.Core.Models;
using NoctusExplorer.Core.Services;
using FluentAssertions;

namespace NoctusExplorer.Core.Tests.Services;

public class KeyBindingResolverTests
{
    [Fact]
    public void Resolve_RegisteredBinding_ReturnsCommandId()
    {
        var resolver = new KeyBindingResolver();
        resolver.SetBinding("pane.copyToOther", new KeyChord("F5"));
        resolver.Resolve(new KeyChord("F5")).Should().Be("pane.copyToOther");
    }

    [Fact]
    public void Resolve_UnregisteredChord_ReturnsNull()
    {
        var resolver = new KeyBindingResolver();
        resolver.Resolve(new KeyChord("F12")).Should().BeNull();
    }

    [Fact]
    public void SetBinding_OverridesPrevious()
    {
        var resolver = new KeyBindingResolver();
        resolver.SetBinding("old.command", new KeyChord("F5"));
        resolver.SetBinding("new.command", new KeyChord("F5"));
        resolver.Resolve(new KeyChord("F5")).Should().Be("new.command");
    }

    [Fact]
    public void GetBinding_ReturnsChordForCommand()
    {
        var resolver = new KeyBindingResolver();
        var chord = new KeyChord("P", ctrl: true, shift: true);
        resolver.SetBinding("tools.commandPalette", chord);
        resolver.GetBinding("tools.commandPalette").Should().Be(chord);
    }

    [Fact]
    public void GetBinding_UnknownCommand_ReturnsNull()
    {
        var resolver = new KeyBindingResolver();
        resolver.GetBinding("nope").Should().BeNull();
    }

    [Fact]
    public void LoadBindings_PopulatesAll()
    {
        var resolver = new KeyBindingResolver();
        resolver.LoadBindings(new Dictionary<string, KeyChord>
        {
            ["file.newTab"] = new KeyChord("T", ctrl: true),
            ["file.closeTab"] = new KeyChord("W", ctrl: true)
        });
        resolver.Resolve(new KeyChord("T", ctrl: true)).Should().Be("file.newTab");
        resolver.Resolve(new KeyChord("W", ctrl: true)).Should().Be("file.closeTab");
    }

    [Fact]
    public void LoadBindings_ClearsPrevious()
    {
        var resolver = new KeyBindingResolver();
        resolver.SetBinding("old", new KeyChord("F1"));
        resolver.LoadBindings(new Dictionary<string, KeyChord>
        {
            ["new"] = new KeyChord("F2")
        });
        resolver.Resolve(new KeyChord("F1")).Should().BeNull();
        resolver.Resolve(new KeyChord("F2")).Should().Be("new");
    }

    [Fact]
    public void RemoveBinding_RemovesMapping()
    {
        var resolver = new KeyBindingResolver();
        resolver.SetBinding("cmd", new KeyChord("F5"));
        resolver.RemoveBinding("cmd");
        resolver.Resolve(new KeyChord("F5")).Should().BeNull();
        resolver.GetBinding("cmd").Should().BeNull();
    }
}
